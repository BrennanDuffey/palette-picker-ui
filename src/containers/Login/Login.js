import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setUser, setProjects, setPalettes } from '../../actions';
import { Redirect, Link } from 'react-router-dom';

class Login extends Component {
  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      error: ''
    };
  };

  handleChange = (e) => {
    this.setState({ [e.target.name]:e.target.value })
  };

  getUser = async (e) => {
    e.preventDefault();
    const { username, password } = this.state;
    try {
      const response = await fetch(`http://localhost:3001/api/v1/users/${username}/${password}`);
      if(!response.ok) {
        throw Error('incorrect username/password');
      }
      const result = await response.json();
      this.props.setUser(result);
    } catch (error) {
      this.setState({ error: error.message });
    };
    await this.getProjects();
  };

  getProjects = async () => {
    try {
      const response = await fetch(`http://localhost:3001/api/v1/projects?user_id=${this.props.user.id}`);
      if(!response.ok) {
        throw Error('Failed to grab projects');
      }
      const projects = await response.json();
      this.props.setProjects(projects);
      this.getPalettes(projects);
    } catch (error) {
      console.log({ error: error.message });
    };
  };

  getPalettes = (projects) => {
    let palettes = [];
    projects.forEach(async (project) => {
      try {
        const response = await fetch(`http://localhost:3001/api/v1/palettes?project_id=${project.id}`);
        const result = await response.json();
        palettes = [...palettes, ...result];
        this.props.setPalettes(palettes);
      } catch(error) {
        console.log({ error: error.message });
      };

    });    
  };

  render() {
    if(this.props.user.username) {
      return <Redirect to='/home' />
    }
    return (
      <form onSubmit={this.getUser} className="login-form">
        <h3>Login</h3>
        <label htmlFor="username">USERNAME:</label>
        <input onChange={this.handleChange} name="username" />
        <label htmlFor="password">PASSWORD:</label>
        <input onChange={this.handleChange} name="password"/> 
        <button type="submit"> Login </button>
        {this.state.error && <p>{this.state.error}</p>}
        <p>
          Not a member? <Link to="/register">Create an account?</Link>
        </p>
      </form>
    );
  };
};

export const mapStateToProps = (state) => ({
  user: state.user,
  projects: state.projects
});

export const mapDispatchToProps = (dispatch) => ({
  setUser: (user) => dispatch(setUser(user)),
  setProjects: (projects) => dispatch(setProjects(projects)),
  setPalettes: (palettes) => dispatch(setPalettes(palettes))
});



export default connect(mapStateToProps, mapDispatchToProps)(Login);