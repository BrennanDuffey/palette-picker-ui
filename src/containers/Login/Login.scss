.Login,
.SavePalette {
  background: $lightGray;
  height: 100vh;
  padding: 20px 0;
  position: relative;
  width: 100%;
  z-index: 500;
}

.login-form,
.save-form {
  background: white;
  display: flex;
  flex-direction: column;
  margin: 50px auto;
  padding: 20px;
  width: 500px;
}

h3 {
  font-size: 2em;
  text-align: center;
}

input,
button {
  font-size: 1.5em;
  margin: 0 0 15px 0;
  padding: 15px;
}

button {
  background: $gray;
  border: none;
  color: white;
  font-weight: 800;

  &:hover {
    background: $lightGray;
  }
}

.login-btn {
  margin: 20px 0 0 0;
}

.signup-msg {
  margin: 5px 0 15px;
}

.signup-link {
  font-size: 1em;
  font-style: italic;
  margin: 0;
}