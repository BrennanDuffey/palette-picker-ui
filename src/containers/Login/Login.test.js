import React from 'react';
import { shallow } from 'enzyme';
import { Login } from './Login';

describe('Login', () => {
  let wrapper;
  let mockUser;

  beforeEach(() => {
    mockUser = {
      id: 2,
      username: 'something'
    };
    wrapper = shallow(
      <Login user={mockUser} />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot on conditional rendering', () => {
    mockUser = {};
    wrapper = shallow(
      <Login user={mockUser}/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    let defaultState = {
      username: '',
      password: '',
      error: ''
    };
    expect(wrapper.state()).toEqual(defaultState);
  });
});