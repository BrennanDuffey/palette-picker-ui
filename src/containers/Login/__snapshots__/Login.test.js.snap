// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<Redirect
  to="/home"
/>
`;

exports[`Login should match snapshot on conditional rendering 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <h3>
    Login
  </h3>
  <label
    htmlFor="username"
  >
    USERNAME:
  </label>
  <input
    name="username"
    onChange={[Function]}
  />
  <label
    htmlFor="password"
  >
    PASSWORD:
  </label>
  <input
    name="password"
    onChange={[Function]}
  />
  <button
    type="submit"
  >
     Login 
  </button>
  <p>
    Not a member? 
    <Link
      to="/register"
    >
      Create an account?
    </Link>
  </p>
</form>
`;
